src/code_builder.rs

```diff
@@ -629,24 +666,26 @@ fn r#return(_: &mut Unstructured, _: &Module, builder: &mut CodeBuilder) -> Resu
     Ok(Instruction::Return)
 }
 
-fn call_valid(module: &Module, builder: &mut CodeBuilder) -> bool {
-    module
-        .funcs()
-        .any(|(_, ty)| builder.types_on_stack(&ty.params))
+fn call_valid(_: &Module, builder: &mut CodeBuilder) -> bool {
+    builder
+        .allocs
+        .functions
+        .keys()
+        .any(|k| builder.types_on_stack(k))
 }
 
 fn call(u: &mut Unstructured, module: &Module, builder: &mut CodeBuilder) -> Result<Instruction> {
-    let n = module
-        .funcs()
-        .filter(|(_, ty)| builder.types_on_stack(&ty.params))
-        .count();
-    debug_assert!(n > 0);
-    let i = u.int_in_range(0..=n - 1)?;
-    let (func_idx, ty) = module
-        .funcs()
-        .filter(|(_, ty)| builder.types_on_stack(&ty.params))
-        .nth(i)
-        .unwrap();
+    let mut candidates = builder
+        .allocs
+        .functions
+        .iter()
+        .filter(|(k, _)| builder.types_on_stack(k))
+        .flat_map(|(_, v)| v.iter().copied())
+        .collect::<Vec<_>>();
+    candidates.sort(); // don't rely on hash map ordering
```

Comment:
I think it would be less of a footgun if we just used a btree map so that we don't have to remember to sort every time we use this stuff.